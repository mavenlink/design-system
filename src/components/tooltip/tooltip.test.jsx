import React from 'react';
import user from '@testing-library/user-event';
import { render, screen } from '@testing-library/react';
import Tooltip from './tooltip.jsx';

describe('tooltip', () => {
  const requiredProps = {
    children: <div id="my-component" aria-describedby="my-component-tooltip">Child</div>,
    id: 'my-component-tooltip',
    text: 'Help Text',
  };

  it('has defaults', () => {
    render(<Tooltip {...requiredProps} />);
    expect(document.body).toMatchSnapshot();

    user.hover(screen.getByText('Child'));
    expect(document.body).toMatchSnapshot();
  });

  it('errors if the tooltip is not describing any elements', () => {
    expect(() => {
      render(
        <Tooltip {...requiredProps}>
          <input />
        </Tooltip>,
      );
    }).toThrowError('<Tooltip> was used without an element on the DOM being described by it. ' +
      'Please add `aria-describedby` to the element this tooltip is being used to describe.');
  });

  describe('hovering behavior', () => {
    it('does not render the tooltip unless the child is hovered', () => {
      render(<Tooltip {...requiredProps} />);
      expect(screen.getByText('Child')).not.toHaveDescription('Help Text');
    });

    it('applies a description to the children when hovered', () => {
      render(<Tooltip {...requiredProps} />);
      user.hover(screen.getByText('Child'));
      expect(screen.getByText('Child')).toHaveDescription('Help Text');
    });

    it('removes the description when the child is no longer hovered', () => {
      render(<Tooltip {...requiredProps} />);

      user.hover(screen.getByText('Child'));
      expect(screen.getByText('Child')).toHaveDescription('Help Text');

      user.unhover(screen.getByText('Child'));
      expect(screen.getByText('Child')).not.toHaveDescription('Help Text');
    });
  });

  describe('focus behavior', () => {
    const children = <input aria-describedby="my-component-tooltip" id="my-component" />;

    it('applies a description to the children when focused', () => {
      render(<Tooltip {...requiredProps}>{children}</Tooltip>);
      user.tab();
      expect(screen.getByRole('textbox')).toHaveFocus();
      expect(screen.getByRole('textbox')).toHaveDescription('Help Text');
    });

    it('removes the description when the child is no longer focused', () => {
      render(<Tooltip {...requiredProps}>{children}</Tooltip>);
      user.tab();
      expect(screen.getByRole('textbox')).toHaveFocus();
      expect(screen.getByRole('textbox')).toHaveDescription('Help Text');

      user.tab();
      expect(screen.getByRole('textbox')).not.toHaveFocus();
      expect(screen.getByRole('textbox')).not.toHaveDescription('Help Text');
    });
  });

  describe('disabled api', () => {
    it('does not show on hover when disabled', () => {
      render(<Tooltip {...requiredProps} disabled={true} />);
      user.hover(screen.getByText('Child'));
      expect(screen.getByText('Child')).not.toHaveDescription('Help Texts');
    });

    it('does not show on focus when disabled', () => {
      render(
        <Tooltip {...requiredProps} disabled={true}>
          <input id="my-component" />
        </Tooltip>,
      );

      user.tab();
      expect(screen.getByRole('textbox')).not.toHaveDescription('Help Texts');
    });
  });

  describe('direction api', () => {
    it('allows top', () => {
      render(<Tooltip {...requiredProps} direction="top" />);
      user.hover(screen.getByText('Child'));
      expect(document.getElementById(requiredProps.id)).toMatchSnapshot();
    });

    it('allows bottom', () => {
      render(<Tooltip {...requiredProps} direction="bottom" />);
      user.hover(screen.getByText('Child'));
      expect(document.getElementById(requiredProps.id)).toMatchSnapshot();
    });

    it('allows left', () => {
      render(<Tooltip {...requiredProps} direction="left" />);
      user.hover(screen.getByText('Child'));
      expect(document.getElementById(requiredProps.id)).toMatchSnapshot();
    });

    it('allows right', () => {
      render(<Tooltip {...requiredProps} direction="right" />);
      user.hover(screen.getByText('Child'));
      expect(document.getElementById(requiredProps.id)).toMatchSnapshot();
    });
  });

  describe('truncate api', () => {
    it('adds the class if true', () => {
      render(<Tooltip {...requiredProps} truncate={true} />);
      user.hover(screen.getByText('Child'));
      expect(document.getElementById(requiredProps.id)).toMatchSnapshot();
    });

    it('does not add the class if false', () => {
      render(<Tooltip {...requiredProps} truncate={false} />);
      user.hover(screen.getByText('Child'));
      expect(document.getElementById(requiredProps.id)).toMatchSnapshot();
    });
  });

  describe('className api', () => {
    it('can be overridden', () => {
      render(<Tooltip {...requiredProps} className="my-dope-class" />);
      user.hover(screen.getByText('Child'));
      expect(document.getElementById(requiredProps.id)).toMatchSnapshot();
    });
  });
});
