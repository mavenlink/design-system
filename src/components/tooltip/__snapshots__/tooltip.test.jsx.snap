// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tooltip className api can be overridden 1`] = `
<span
  class="my-dope-class top"
  id="my-component-tooltip"
  style="left: 0px; top: -128px;"
>
  Help Text
</span>
`;

exports[`tooltip direction api allows bottom 1`] = `
<span
  class="tooltip bottom"
  id="my-component-tooltip"
  style="left: 0px; top: 128px;"
>
  Help Text
</span>
`;

exports[`tooltip direction api allows left 1`] = `
<span
  class="tooltip left"
  id="my-component-tooltip"
  style="left: -128px; top: 0px;"
>
  Help Text
</span>
`;

exports[`tooltip direction api allows right 1`] = `
<span
  class="tooltip right"
  id="my-component-tooltip"
  style="left: 128px; top: 0px;"
>
  Help Text
</span>
`;

exports[`tooltip direction api allows top 1`] = `
<span
  class="tooltip top"
  id="my-component-tooltip"
  style="left: 0px; top: -128px;"
>
  Help Text
</span>
`;

exports[`tooltip has defaults 1`] = `
<body>
  <div>
    <div
      class="wrapper"
    >
      <div
        aria-describedby="my-component-tooltip"
        id="my-component"
      >
        Child
      </div>
    </div>
  </div>
</body>
`;

exports[`tooltip has defaults 2`] = `
<body>
  <div>
    <div
      class="wrapper"
    >
      <div
        aria-describedby="my-component-tooltip"
        id="my-component"
      >
        Child
      </div>
    </div>
  </div>
  <span
    class="tooltip top"
    id="my-component-tooltip"
    style="left: 0px; top: -128px;"
  >
    Help Text
  </span>
</body>
`;

exports[`tooltip truncate api adds the class if true 1`] = `
<span
  class="tooltip top truncate"
  id="my-component-tooltip"
  style="left: 0px; top: -128px;"
>
  Help Text
</span>
`;

exports[`tooltip truncate api does not add the class if false 1`] = `
<span
  class="tooltip top"
  id="my-component-tooltip"
  style="left: 0px; top: -128px;"
>
  Help Text
</span>
`;
